src/transformation/mesh_intersection/mesh_intersection.rs

```diff
@@ -256,393 +348,547 @@ fn extract_connected_components(
+fn smallest_angle(points: &[Point3<Real>]) -> Real {
+    let n = points.len();
+
+    let mut worst_cos = -2.0;
+    for i in 0..points.len() {
+        let d1 = (points[i].coords - points[(i + 1) % n].coords).normalize();
+        let d2 = (points[(i + 2) % n].coords - points[(i + 1) % n].coords).normalize();
+
+        let cos = d1.dot(&d2);
+        if cos > worst_cos {
+            worst_cos = cos;
         }
-        FeatureId::Vertex(vid) => FeatureId::Vertex(mesh.indices()[tri as usize][vid as usize]),
-        FeatureId::Face(_) => FeatureId::Face(tri),
-        FeatureId::Unknown => FeatureId::Unknown,
     }
+
+    worst_cos.acos()
 }
 
-fn unified_vertex(
-    mesh1: &TriMesh,
-    mesh2: &TriMesh,
-    new_vertices12: &[Point<Real>],
-    pos12: &Isometry<Real>,
-    vid: u32,
-) -> Point<Real> {
-    let base_id2 = mesh1.vertices().len() as u32;
-    let base_id12 = (mesh1.vertices().len() + mesh2.vertices().len()) as u32;
-
-    if vid < base_id2 {
-        mesh1.vertices()[vid as usize]
-    } else if vid < base_id12 {
-        pos12 * mesh2.vertices()[(vid - base_id2) as usize]
-    } else {
-        new_vertices12[(vid - base_id12) as usize]
+fn planar_gram_schmidt(v1: Vector3<Real>, v2: Vector3<Real>) -> (Vector3<Real>, Vector3<Real>) {
+    let u1 = v1;
+    let u2 = v2 - (v2.dot(&u1) / u1.norm_squared()) * u1;
+
+    let e1 = u1.normalize();
+    let e2 = u2.normalize();
+
+    (e1, e2)
+}
+
+fn project_point_to_segment(point: &Vector3<Real>, segment: &[Vector3<Real>; 2]) -> Vector3<Real> {
+    let dir = segment[1] - segment[0];
+    let local = point - segment[0];
+
+    let norm = dir.norm();
+    // Restrict the result to the segment portion of the line.
+    let coeff = (dir.dot(&local) / norm).clamp(0., norm);
+
+    segment[0] + coeff * dir.normalize()
+}
+
+/// No matter how smart we are about computing intersections. It is always possible
+/// to create ultra thin triangles when a point lies on an edge of a tirangle. These
+/// are degenerate and need to be terminated with extreme prejudice.
+fn is_triangle_degenerate(
+    triangle: &[Point3<Real>; 3],
+    epsilon_angle: Real,
+    epsilon_distance: Real,
+) -> bool {
+    if smallest_angle(triangle) < epsilon_angle {
+        return true;
+    }
+
+    let mut shortest_side = Real::MAX;
+    for i in 0..3 {
+        let p1 = triangle[i];
+        let p2 = triangle[(i + 1) % 3];
+
+        shortest_side = shortest_side.min((p1 - p2).norm());
     }
+
+    let mut worse_projection_distance = Real::MAX;
+    for i in 0..3 {
+        let dir = triangle[(i + 1) % 3] - triangle[(i + 2) % 3];
+        if dir.norm() < epsilon_distance {
+            return true;
+        }
+
+        let dir = dir.normalize();
+        let proj =
+            (triangle[i] - triangle[(i + 2) % 3]).dot(&dir) * dir + triangle[(i + 2) % 3].coords;
+
+        worse_projection_distance =
+            worse_projection_distance.min((proj - triangle[i].coords).norm());
```

Comment:
If the suggestion above is applied, this `.coords` isn’t needed either.
```suggestion
            worse_projection_distance.min((proj - triangle[i]).norm());
```